controllers/openstackdataplanenode_controller.go

```diff
@@ -166,3 +183,71 @@ func (r *OpenStackDataPlaneNodeReconciler) ConfigureNetwork(ctx context.Context,
 
 	return nil
 }
+
+// GetInstanceRole returns the role of a node based on the node's role name
+func (r *OpenStackDataPlaneNodeReconciler) GetInstanceRole(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode) (*dataplanev1beta1.OpenStackDataPlaneRole, error) {
+	// Use the instances's role name to get its role object
+	var err error
+	instanceRole := &dataplanev1beta1.OpenStackDataPlaneRole{}
+	err = r.Client.Get(ctx, client.ObjectKey{
+		Namespace: instance.Namespace,
+		Name:      instance.Spec.Role,
+	}, instanceRole)
+	return instanceRole, err
+}
+
+// GetAnsibleUser retruns the string value from the template unless it is set in the node
+func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleUser(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {
+	if instance.Spec.Node.AnsibleUser != "" {
+		return instance.Spec.Node.AnsibleUser
+	}
+	return instanceRole.Spec.NodeTemplate.AnsibleUser
+}
+
+// GetAnsiblePort retruns the string value from the template unless it is set in the node
+func (r *OpenStackDataPlaneNodeReconciler) GetAnsiblePort(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {
+	if instance.Spec.Node.AnsiblePort > 0 {
+		return strconv.Itoa(instance.Spec.Node.AnsiblePort)
+	}
+	return strconv.Itoa(instanceRole.Spec.NodeTemplate.AnsiblePort)
```

Comment:
Should we raise an error when fields are empty in both node and role?
```suggestion
	if instanceRole.Spec.NodeTemplate.AnsiblePort > 0 {
		return strconv.Itoa(instanceRole.Spec.NodeTemplate.AnsiblePort), nil
	}
	return nil, fmt.Errorf("Please provide the AnsiblePort")
```

Reply:
I don't disagree about trying to capture that type of error. However, I think we should do it here in the controller and instead do it as a validation in the API.

https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/api/v1beta1/openstackdataplanenode_types.go#L70

Reply:
I agree with @fultonj that we might want to do that check elsewhere.
I think we don't have to check if the value is `> 0`: in the API spec [1] we can use `kubebuilder` to validate the provided input, which can be set as `22` as default.

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/api/v1beta1/openstackdataplanenode_types.go#L70

Reply:
I'm doing this check (>0) because I want to know if the value has been set to something other than the default. 

I was thinking of doing a separate PR to add the validation and set a new default of 22. Though, if the ansible port is not specified in an inventory, I think ansible chooses 22. Thus, we don't want to always set it to 22 in the inventory. We only want to set it in the inventory if it's non standard. Otherwise it will just be noise. We can add a PR with the validation and continue this discussion there.